src/runtime/recipe/walker.ts

```diff
@@ -5,126 +5,211 @@
-  // tslint:disable-next-line: no-any
-  onSlot?(recipe: Recipe, slot: Slot): any;
-  // tslint:disable-next-line: no-any
-  onObligation?(recipe: Recipe, obligation: ConnectionConstraint): any;
-
-  onResult(result) {
-    super.onResult(result);
-    const recipe: Recipe = result.result;
-    const updateList = [];
-
-    // update phase - walk through recipe and call onRecipe,
-    // onHandle, etc.
-
-    if (this.onRecipe) {
-      result = this.onRecipe(recipe, result);
-      if (!this.isEmptyResult(result)) {
-        updateList.push({continuation: result});
-      }
-    }
-    for (const particle of recipe.particles) {
-      if (this.onParticle) {
-        const context: [Particle] = [particle];
-        const result = this.onParticle(recipe, ...context);
-        if (!this.isEmptyResult(result)) {
-          updateList.push({continuation: result, context});
-        }
-      }
-    }
-    for (const handleConnection of recipe.handleConnections) {
-      if (this.onHandleConnection) {
-        const context: [HandleConnection] = [handleConnection];
-        const result = this.onHandleConnection(recipe, ...context);
-        if (!this.isEmptyResult(result)) {
-          updateList.push({continuation: result, context});
-        }
-      }
-    }
-    for (const handle of recipe.handles) {
-      if (this.onHandle) {
-        const context: [Handle] = [handle];
-        const result = this.onHandle(recipe, ...context);
-        if (!this.isEmptyResult(result)) {
-          updateList.push({continuation: result, context});
-        }
-      }
-    }
-    if (this.onPotentialSlotConnection) {
-      for (const particle of recipe.particles) {
-        for (const [name, slotSpec] of particle.getSlotSpecs()) {
-          if (particle.getSlotConnectionByName(name)) continue;
-          const context: [Particle, SlotSpec] = [particle, slotSpec];
-          const result = this.onPotentialSlotConnection(recipe, ...context);
-          if (!this.isEmptyResult(result)) {
-            updateList.push({continuation: result, context});
+  static Independent: WalkerTactic = WalkerTactic.Independent;
+  descendants: Descendant[];
+  currentAction;
+  currentResult;
+  tactic: WalkerTactic;
+
+  constructor(tactic) {
+    this.descendants = [];
+    assert(tactic);
+    this.tactic = tactic;
+  }
+
+  onAction(action: Action) {
+    this.currentAction = action;
+  }
+
+  onResult(result): void {
+    this.currentResult = result;
+  }
+
+  onResultDone(): void {
+    this.currentResult = undefined;
+  }
+
+  onActionDone(): void {
+    this.currentAction = undefined;
+  }
+
+  static walk(results, walker: Walker, action: Action): Descendant[] {
+    walker.onAction(action);
+    results.forEach(result => {
+      walker.onResult(result);
+      walker.onResultDone();
+    });
+    walker.onActionDone();
+    return walker.descendants;
+  }
+
+  _runUpdateList(recipe, updateList) {
```

Comment:
Would you care about renaming the arguments and variables below to not mention _recipes_, as walker is supposed to be a generic base class?

Reply:
Good point. This required a bit more refactoring, so PTAL.